### PR TITLE
#80 codeflow thread support

### PR DESCRIPTION
**Pull request by EasyRhin0**

### CODE REVIEW

**Comment by EasyRhin0** [comment]
The SDV converter is the only one that has call/return (unless I'm missing something). In that converter, nestingLevel is set on lines 202 and 226. Please check one of the test logs under Sarif.FunctionalTests\ConverterTestData\StaticDriverVerifier and see if it looks right..

**Comment by lgolding** [comment]
@EasyRhin0 You're right that the SDV converter is the only one that needs to deal with this "cross-functional" flows. As I mentioned in my comment above, I looked at the very first SDV log and it did not have `nestingLevel` populated.

File: src/Sarif.FunctionalTests/SpecExamples/Comprehensive.sarif [hunk 1]
```diff
@@ -220,76 +220,76 @@
...
-                    "physicalLocation": {
-                      "fileLocation": {
-                        "uri": "file:///home/buildAgent/src/collections/list.h"
-                      },
-                      "region": {
-                        "snippet": {
-                          "text": "int *ptr;"
+                  "locations": [
+                    {
+                      "step": 0,
```
[the remark this answers is not in the document]
Comment (EasyRhin0) [reply src/Sarif.FunctionalTests/SpecExamples/Comprehensive.sarif]: Fixed

**Comment by lgolding** [comment]
(See [cancelspinlock_bug1.tt.sarif](https://raw.githubusercontent.com/Microsoft/sarif-sdk/75362b74cca0dda9a041bf60e0d6e13b6be0fa5c/src/Sarif.FunctionalTests/ConverterTestData/StaticDriverVerifier/cancelspinlock_bug1.tt.sarif) -- no mention of `nestingLevel`.)

File: src/Sarif.Converters/StaticDriverVerifierConverter.cs [hunk 2]
```diff
@@ -91,14 +82,14 @@ private Result ProcessSdvDefectStream(Stream input)
 
                 while (!string.IsNullOrEmpty(line = reader.ReadLine()))
                 {
-                    ProcessLine(line, nestingLevel, result);
+                    ProcessLine(line, ref nestingLevel, result);
```
Comment (lgolding) [inline src/Sarif.Converters/StaticDriverVerifierConverter.cs]: :-)

File: src/Sarif.Converters/CppCheckError.cs [hunk 3]
```diff
@@ -158,16 +158,7 @@ public Result ToSarifIssue()
...
-                    {
-                        ThreadFlows = new List<ThreadFlow>()
-                        {
-                            new ThreadFlow
-                            {
-                                Locations = locations
-                            }
-                        }
-                    }
+                    SarifUtilities.CreateSingleThreadedCodeFlow(locations)
```
Comment (lgolding) [inline src/Sarif.Converters/CppCheckError.cs]: How lovely! Not just shorter, but the name communicates.

File: src/Sarif.FunctionalTests/ConverterTestData/StaticDriverVerifier/cancelspinlock_bug1.tt.sarif [hunk 4]
```diff
@@ -52,13 +52,15 @@
                     },
                     {
                       "step": 1,
```
Comment (lgolding) [inline src/Sarif.FunctionalTests/ConverterTestData/StaticDriverVerifier/cancelspinlock_bug1.tt.sarif]: I see that the steps are still 0-based (note that the first location in this flow has no "step" property, so it's 0). Is there another commit coming?

File: src/Sarif.Converters/FortifyFprConverter.cs [hunk 5]
```diff
@@ -345,7 +345,7 @@ private void ParseLocationFromAnalysisInfo(Result result)
 
                     var cfl = new CodeFlowLocation
                     {
-                        Step = step++,
+                        Step = ++step,
```
Comment (lgolding) [inline src/Sarif.Converters/FortifyFprConverter.cs]: Lovely :-)

**Pull request merged**